# TD/TP 1 : Démarrer en python

### 1. Types de variables

**1.1** Les instructions suivantes créent et initialisent les variables n, p et x. Quels sont les types de ces trois variables ?

    n = 10 
    p = 7 
    x = 2. 

**1.2** Compléter et exécuter les cellules ci-dessous de façon à vérifier vos réponses.

In [5]:
import numpy as np 

In [2]:
# instruction pour créer n 
n = 10 
# instruction pour afficher la valeur de n
print(n)
# instruction pour renvoyer le type de n
type(n)

10


int

In [3]:
# instruction pour créer p
p = 7
# instruction pour afficher la valeur de p
print(p)
# instruction pour renvoyer le type de p
type(p)


7


int

In [4]:
# instruction pour créer x 
x = 2
# instruction pour afficher la valeur de x
print(x) 
# instruction pour renvoyer le type de x
type(x)


2


int

**1.3** Donner le type et la valeur des expressions suivantes :

       a. x+n/p
       b. (x+n)/p
       c. 5. * n
       d. (n + 1) / n
       e. (n + 1) // n
       f. (n + 1.) // n
    
Vérifiez vos résultats en exécutant les instructions dans la cellule ci-dessous.

In [5]:
# instructions pour vérifier les réponses de la question a
a = x+n/p ; print("a=",a,".Type :", type(a)) 

a= 3.428571428571429 .Type : <class 'float'>


In [6]:
# instructions pour vérifier les réponses de la question b
b = (x+n)/p ; print("b=",b,".Type :", type(b) )

b= 1.7142857142857142 .Type : <class 'float'>


In [7]:
# instructions pour vérifier les réponses de la question c
c = 5.*n ;  print("c=",c,".Type :", type(c))

c= 50.0 .Type : <class 'float'>


In [8]:
# instructions pour vérifier les réponses de la question d
d = (n+1) / n ;  print("d=",d,".Type :", type(d))

d= 1.1 .Type : <class 'float'>


In [9]:
# instructions pour vérifier les réponses de la question e
e = (n+1) // n ;  print("e=",e,".Type :", type(e)) # Division entiere

e= 1 .Type : <class 'int'>


In [10]:
# instructions pour vérifier les réponses de la question f
f = (n+1.) // n ;  print("f=",f,".Type :", type(f)) 

f= 1.0 .Type : <class 'float'>


### 2. Entrées / sorties

**2.1** Ecrire un programme qui demande à l’utilisateur un entier puis l’affiche. Un exemple d’éxécution est donné ci-dessous :

    Veuillez entrer un entier : 42
    Vous avez saisi 42

In [13]:
# bloc d'instructions à compléter et à exécuter
x  = int(input("veuillez saisir un entier :"))
print("vous avez saisi",x)

vous avez saisi 12


**2.2** Compléter le programme précédent pour qu’il demande aussi un réel puis affiche les 2 valeurs saisies. Un exemple d’exécution est donné ci-dessous :

    Veuillez entrer un entier : 42
    Veuillez entrer un réel : 13.37
    Vous avez saisi 42 et 13.37

In [14]:
# bloc d'instructions à compléter et à exécuter
x  = int(input("veuillez saisir un entier :"))
y = float(input("veuillez saisir un réel :"))
print("vous avez saisi",x,"et",y)

vous avez saisi 12 et 13.77


**2.3** Ecrire un programme qui demande à l’utilisateur un charactère puis affiche le caractère et son code ASCII. On rappelle que le code ASCII d’une lettre correspond à un entier donnant sa position dans la table ASCII ; on pourra l’obtenir par l’appel à la fonction ord. Un exemple d’exécution est donné ci-dessous :

    Veuillez entrer un charactère : b
    Vous avez saisi b dont le code ASCII est 98

In [15]:
# bloc d'instructions à compléter et à exécuter
char= str(input("Veuillez saisir un charactère :"))
value = ord(char)
print("Vous avez saisi",char,"dont le code ASCII est",value)

Vous avez saisi x dont le code ASCII est 120


### 3. Structures de contrôle

**3.1** Ecrire un programme qui demande un nombre à l'utilisateur et teste si ce nombre est positif.

In [7]:
# bloc d'instructions à compléter et à exécuter
n = int(input("Saisir un nombre :"))
if np.sign(n) == + 1 : 
    print("Le nombre saisi est positif")
else : 
    print("il ne l'est pas")



il ne l'est pas


**3.2** Ecrire un programme qui demande un nombre à l’utilisateur et teste si ce nombre est strictement positif, nul
ou strictement négatif.

In [17]:
# bloc d'instructions à compléter et à exécuter
n = int(input("Saisir un nombre :"))
if np.sign(n) == + 1 : 
    print("Le nombre saisi est positif")
if np.sign(n) == - 1 : 
    print("Le nombre saisi est négatif")
else : 
    print("Le nombre saisi est nul")



Le nombre saisi est négatif


**3.3** Ecrire un programme qui demande une année et affiche si celle-ci est bissextile ou non. Les années bissextiles
sont les années : soit divisibles par 4 mais non divisibles par 100, soit divisibles par 400.

In [18]:
# bloc d'instructions à compléter et à exécuter
year = int(input("Saisir une année :"))
a = year % 4 ; b = year % 100 ; c = year % 400 
if (a==0 and (b!=0)) or c==0 : 
    print("L'année",year,"est bissextile")
else : print("L'année", year , "n'est pas bissextile")

L'année 1936 est bissextile


### 4. Fonctions, boucles et premiers algorithmes

**4.1** Ecrire une _fonction_ qui calcule et renvoie la somme des N premiers entiers. N est un paramètre d'entrée de la fonction. Tester la fonction sur quelques exemples. Un exemple d’éxécution est donné ci-dessous :

    Veuillez entrer un entier : 5
    La somme des 5 premiers entiers est 15

In [20]:
# bloc d'instructions à compléter et à exécuter
""""def Sum(N) : 
    
    S = 0 # Initialisation 
    for k in range(1,N+1) : # k allant de 1 à N 
        S += k  # A chaque tour , on somme les indices 
    
    return S 

N = int(input("Veuillez entrer un entier:"))
Sum(N)"""



# Une deuxieme approche : 
def Somme(K) : 
    Somme = [i for i in range(K+1)] # Tableau de valeurs allant de 0 à K 
    Summ = sum(Somme) # Fonction de sommation 
    
    return Summ 

K = int(input("Veuillez entrer un entier:"))
print("La somme des des",K,"premiers entiers est",Somme(K))





21

**4.2** Ecrire une _fonction_ qui calcule et renvoie N!, où N est un paramètre d'entrée de la fonction. Tester la fonction sur quelques exemples. Un exemple d’éxécution est donné ci-dessous :

    Veuillez entrer un entier : 5
    5! = 120

In [8]:
# bloc d'instructions à compléter et à exécuter
def Fact(N) : 
    F = 1 
    for i in range (1,N+1) : # de 1 à N 
        F *= i    
        
    return F
    

K = int(input("Veuillez entrer un entier:"))
print(K,"! vaut:",Fact(K))



7 ! vaut: 5040


**4.3** Ecrire une _fonction_ qui calcule et renvoie x^N, où x et N sont des paramètres d'entrée de la fonction. Tester la fonction sur quelques exemples. Un exemple d’éxécution est donné ci-dessous :

    Veuillez entrer un réel : 2.5
    Veuillez entrer un entier positif : 3
    2.5^3 = 15.625

In [24]:
# bloc d'instructions à compléter et à exécuter
def Exposant(x,N) : 
    return x**N

x = float(input("Veuillez enter un réel :"))
N = int(input("Veuillez enter un entier positif :"))
print(x,"^",N,"=",Exposant(x,N))

7.0 ^ 3 = 343.0


**4.4** Ecrire une _fonction_ qui utilise le développement en série de exp pour calculer une valeur approchée de exp(x), avec une précision donnée eps = 10^(-n). x et n sont les paramètres d'entrée de la fonction. La fonction renvoie la valeur de exp(x) ainsi que le nombre de termes du développement en série utilisé pour la précision considérée. Tester la fonction sur quelques exemples. Vérifier vos résultats avec la fonction native _exp_. Un exemple d’éxécution est donné ci-dessous :

    Calcul de exp(x) à 10^-n près :
    Veuillez entrer x : 2.
    Veuillez entrer n : 3 
    Résultat : exp(2.0) = 
    exp(2.0) = 7.3887125220458545 - calcul avec 10 terme(s)
    Vérification : math.exp(x) = 7.38905609893065

In [3]:
# bloc d'instructions à compléter et à exécuter

# Definition du DL en 0 de l'exp 

"""def Exp(x,n) : 
    k  = [ i for i in range(n+1)] # k va de 0 à N 
    W = 0
    for j in k : 
        W += Exposant(x,k[j]) / Fact(k[j])
        
    return W


Exp(3,3)""" 

# Correction 
def exponentielle(x,n) :
    eps = 10**(-n) # precision 
    exp = 0
    increment = 1
    k = 0
    while (increment > eps) :
        exp += increment
        k += 1
        increment = increment * x / k
    return exp, k




In [4]:
import math
print('Calcul de exp(x) à 10^-n près :')
x = float(input('Veuillez saisir x : '))
n = int(input('Veuillez saisir n : '))
resultat = exponentielle(x,n)
print('Résultat : exp(',x,') = ',resultat[0])
print('Calcul avec', resultat[1], 'termes')
print('Vérification : math.exp(x) =', math.exp(x))

Calcul de exp(x) à 10^-n près :
Résultat : exp( 3.0 ) =  20.079665178571425
Calcul avec 11 termes
Vérification : math.exp(x) = 20.085536923187668


### 5. Boucles imbriquées

**5.1** Ecrire une fonction qui affiche un échiquier de largeur N, où N est un paramètre d'entrée de la fonction. L'échiquier est un carré de côté N, dessiné avec les conventions suivantes : (a) les cases noires sont représentées par un ”X” et les cases blanches par ” ”, (b) le tour de l’échiquier et de chaque case est représenté par les caractères + (coin), | (ligne verticale) et - (ligne horizontale). 

Tester la fonction sur quelques exemples. Un exemple d’affichage est  montré ci-dessous ( échiquier 4x4) :

    Veuillez entrer un entier : 4
    Voici un échiquier 4x4
    +---+---+---+---+
    |   | X |   | X |
    +---+---+---+---+
    | X |   | X |   |
    +---+---+---+---+
    |   | X |   | X |
    +---+---+---+---+
    | X |   | X |   |
    +---+---+---+---+

In [69]:
# bloc d'instructions à compléter et à exécuter

def Echec(Z) : 
   
   ligne = ('+---')  ; nb  = ('| X |' , '   ') ;  bn = ('|   |', ' X ')
   for i in range(Z+1) : # Parcours de lignes
      print(ligne*Z,'+',sep='') ; nb *= Z ; bn *= Z
      
      for j in range(Z) : # Parcours de cases
        
        if i != Z and i % 2 == 0 : # Ligne paire
           print(bn[j],end='') 
           if Z % 2==0 and j == Z-1 : print('|','') # Fermeture droite du carré pour Z pair 
           elif Z % 2!= 0 and j == Z-1 : print('')
           
        elif  i != Z and i % 2 != 0: # Ligne Impaire 
          print(nb[j],end='') 
          if Z % 2==0 and j == Z-1 : print('|','') # Fermeture droite du carré pour Z pair 
          elif Z % 2!= 0 and j == Z-1 : print('') 

       



Z = int(input("Veuillez entrer un entier : "))   
Echec(Z) ; print("\n")
Echec(Z-1)


+---+---+---+---+---+---+
|   | X |   | X |   | X | 
+---+---+---+---+---+---+
| X |   | X |   | X |   | 
+---+---+---+---+---+---+
|   | X |   | X |   | X | 
+---+---+---+---+---+---+
| X |   | X |   | X |   | 
+---+---+---+---+---+---+
|   | X |   | X |   | X | 
+---+---+---+---+---+---+
| X |   | X |   | X |   | 
+---+---+---+---+---+---+


+---+---+---+---+---+
|   | X |   | X |   |
+---+---+---+---+---+
| X |   | X |   | X |
+---+---+---+---+---+
|   | X |   | X |   |
+---+---+---+---+---+
| X |   | X |   | X |
+---+---+---+---+---+
|   | X |   | X |   |
+---+---+---+---+---+


**5.2** Ecrire une fonction qui affiche une pyramide droite de N lignes composées d'un caractère c, où N et c sont des paramètres d'entrée de la fonction. 
Tester la fonction sur quelques exemples. Un exemple d’affichage est  montré ci-dessous :

    Veuillez entrer un entier : 4
    Veuillez entrer un caractère : h
    Voici la pyramide :
    h
    hh
    hhh
    hhhh

In [25]:
# bloc d'instructions à compléter et à exécuter
def Pyramide(N,c) : 
   
  for i in range(1,N+1) : 
      print(c*i,"\n")
    

N = int(input("Veuillez entrer un entier :"))
c = str(input("Veuillez enter un caractère :"))
Pyramide(N,c)


Q 

QQ 

QQQ 

QQQQ 

QQQQQ 

QQQQQQ 

QQQQQQQ 

